lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -3072,11 +3087,64 @@ ConformanceChecker::inferTypeWitnessesViaValueWitnesses(
           if (!dmt->getBase()->isEqual(Conformance->getType()))
             return false;
           
+          // If this associated type is same-typed to another associated type
+          // on `Self`, then it may still be an interesting candidate if we find
+          // an answer for that other type.
+          if (witness->getDeclContext()->getAsProtocolExtensionContext()
+              && witness->getDeclContext()->getGenericSignatureOfContext()) {
+            auto selfTy = witness->getDeclContext()->getSelfInterfaceType();
+            auto selfAssocTy = DependentMemberType::get(selfTy,
+                                                        dmt->getAssocType());
+            for (auto &reqt : witness->getDeclContext()
```

Comment:
I feel like factoring out 'witness->getDeclContext()' and 'dc->getGenericSignatureOfContext()' would make this code easier to read.